File: packages/provide-uterm-go/hub/presence_test.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package hub

import "testing"

func TestCanSendInputHijackMode(t *testing.T) {
	h, clk := newTestHub(t, nil)
	owner := newBrowserWS("owner")
	other := newBrowserWS("other")
	st := NewWorkerTermState()
	st.Browsers[owner] = "viewer"
	st.Browsers[other] = "operator"
	st.HijackOwner = owner
	exp := clk.Monotonic() + 100
	st.HijackOwnerExpiresAt = &exp

	mustTrue(t, h.Presence.CanSendInput(st, owner), "dashboard owner can send")
	mustFalse(t, h.Presence.CanSendInput(st, other), "non-owner cannot send in hijack mode")
}

func TestCanSendInputOpenMode(t *testing.T) {
	h, _ := newTestHub(t, nil)
	viewer := newBrowserWS("v")
	op := newBrowserWS("o")
	admin := newBrowserWS("a")
	unknown := newBrowserWS("u")
	st := NewWorkerTermState()
	st.InputMode = InputModeOpen
	st.Browsers[viewer] = "viewer"
	st.Browsers[op] = "operator"
	st.Browsers[admin] = "admin"

	mustFalse(t, h.Presence.CanSendInput(st, viewer), "viewer excluded in open mode")
	mustTrue(t, h.Presence.CanSendInput(st, op), "operator may send in open mode")
	mustTrue(t, h.Presence.CanSendInput(st, admin), "admin may send in open mode")
	mustFalse(t, h.Presence.CanSendInput(st, unknown), "unknown ws defaults to viewer -> excluded")
}

func TestRegisterBrowserStateSnapshot(t *testing.T) {
	h, clk := newTestHub(t, nil)
	// Unknown worker returns defaults.
	got := h.Presence.RegisterBrowserStateSnapshot("nope", newBrowserWS("x"))
	mustEqual(t, got["is_hijacked"], false, "unknown is_hijacked")
	mustEqual(t, got["worker_online"], false, "unknown worker_online")
	mustEqual(t, got["input_mode"], "hijack", "unknown input_mode")

	worker := &fakeWorkerWS{}
	ws := newBrowserWS("b")
	st := registerWorkerState(h, "w1", worker)
	st.Browsers[ws] = "operator"
	st.HijackOwner = ws
	exp := clk.Monotonic() + 100
	st.HijackOwnerExpiresAt = &exp

	got = h.Presence.RegisterBrowserStateSnapshot("w1", ws)
	mustEqual(t, got["is_hijacked"], true, "hijacked")
	mustEqual(t, got["hijacked_by_me"], true, "hijacked_by_me")
	mustEqual(t, got["worker_online"], true, "worker_online")
	mustEqual(t, got["input_mode"], "hijack", "input_mode")
}

func TestRequestSnapshotAndAnalysisFrames(t *testing.T) {
	seq := 0
	h, _ := newTestHub(t, func(c *TermHubConfig) {
		c.IDGen = func() string { seq++; return "id" + string(rune('0'+seq)) }
	})
	worker := &fakeWorkerWS{}
	registerWorkerState(h, "w1", worker)

	mustEqual(t, h.Presence.RequestSnapshot(bg(), "w1"), nil, "snapshot req err")
	frame := decodeOneControl(t, worker.last())
	mustEqual(t, frame["type"], "snapshot_req", "snapshot_req type")
	mustEqual(t, frame["req_id"], "id1", "req_id")
	if _, ok := frame["ts"]; !ok {
		t.Fatal("snapshot_req missing ts")
	}

	mustEqual(t, h.Presence.RequestAnalysis(bg(), "w1"), nil, "analysis req err")
	frame = decodeOneControl(t, worker.last())
	mustEqual(t, frame["type"], "analyze_req", "analyze_req type")
	mustEqual(t, frame["req_id"], "id2", "req_id 2")
}

func TestRequestSnapshotNoWorker(t *testing.T) {
	h, _ := newTestHub(t, nil)
	// No worker connected: send_worker returns (false, nil) — no error.
	mustEqual(t, h.Presence.RequestSnapshot(bg(), "ghost"), nil, "no-op snapshot req")
	mustEqual(t, h.Presence.RequestAnalysis(bg(), "ghost"), nil, "no-op analysis req")
}

func TestUUID4Format(t *testing.T) {
	id := newUUID4()
	// 8-4-4-4-12 with version/variant nibbles.
	if len(id) != 36 || id[14] != '4' {
		t.Fatalf("unexpected uuid4 format: %q", id)
	}
}

File: packages/provide-uterm-ts/src/hub/presence.test.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

import { describe, expect, it } from "vitest";
import { loadGolden } from "../testing/golden.ts";
import {
  type BrowserRole,
  type PresenceHubCallbacks,
  PresenceManager,
  WorkerRegistry,
  WorkerTermState,
} from "./index.ts";

interface HubPresenceGolden {
  now: number;
  input: Array<{
    name: string;
    input_mode: "hijack" | "open";
    role: BrowserRole | null;
    is_owner: boolean;
    allowed: boolean;
  }>;
  snapshots: Array<{
    name: string;
    is_hijacked: boolean;
    hijacked_by_me: boolean;
    worker_online: boolean;
    input_mode: string;
  }>;
  control_frames: {
    types: string[];
    timestamps: number[];
    req_id_lengths: number[];
    req_ids_differ: boolean;
    keys: string[][];
  };
  roles: { resolved: BrowserRole[] };
}

const golden = loadGolden<HubPresenceGolden>("hub_presence_golden.json");
const NOW = golden.now;
const ASKER = { id: "asker" };
const OTHER = { id: "someone-else" };

/** A recording stand-in for the hub surface presence reaches through. */
class FakeHub implements PresenceHubCallbacks {
  readonly registry = new WorkerRegistry<WorkerTermState>();
  readonly sent: Array<Record<string, unknown>> = [];
  role: BrowserRole = "viewer";

  isDashboardHijackActive(state: WorkerTermState): boolean {
    if (state.hijackOwner === undefined) {
      return false;
    }
    return state.hijackOwnerExpiresAt === undefined || state.hijackOwnerExpiresAt > NOW;
  }

  hasValidRestLease(state: WorkerTermState): boolean {
    return state.hijackSession !== undefined && state.hijackSession.leaseExpiresAt > NOW;
  }

  isHijacked(state: WorkerTermState): boolean {
    return this.isDashboardHijackActive(state) || this.hasValidRestLease(state);
  }

  async resolveRoleForBrowser(): Promise<BrowserRole> {
    return this.role;
  }

  async sendWorker(_workerId: string, message: Record<string, unknown>): Promise<boolean> {
    this.sent.push(message);
    return true;
  }
}

/** A presence manager over a recording hub. */
function build() {
  const hub = new FakeHub();
  const presence = new PresenceManager({ hub, wallNow: () => NOW });
  return { hub, presence };
}

/** How each recorded snapshot case sets its worker up. */
const SNAPSHOT_SHAPES: Record<
  string,
  | {
      connected?: boolean;
      inputMode?: "hijack" | "open";
      owner?: object;
      ownerExpiresAt?: number;
      restExpiresAt?: number;
    }
  | undefined
> = {
  "unknown worker": undefined,
  "idle worker": {},
  "worker offline": { connected: false },
  "hijacked by me": { owner: ASKER, ownerExpiresAt: NOW + 10 },
  "hijacked by someone else": { owner: OTHER, ownerExpiresAt: NOW + 10 },
  "my hijack has expired": { owner: ASKER, ownerExpiresAt: NOW - 10 },
  "rest lease held": { restExpiresAt: NOW + 10 },
  "open input mode": { inputMode: "open" },
  "perpetual hold by me": { owner: ASKER },
  "my hold lapsed, rest took over": { owner: ASKER, ownerExpiresAt: NOW - 10, restExpiresAt: NOW + 10 },
};

describe("PresenceManager.canSendInput", () => {
  it.each(golden.input)("$name", (record) => {
    // This runs on every browser input frame. In hijack mode only the lease
    // holder may type; in open mode the lease is irrelevant and the role
    // decides, with viewers still refused.
    const { presence } = build();
    const state = new WorkerTermState({ now: () => NOW });
    state.inputMode = record.input_mode;
    state.hijackOwner = record.is_owner ? ASKER : OTHER;
    state.hijackOwnerExpiresAt = NOW + 10;
    if (record.role !== null) {
      state.browsers.set(ASKER, record.role);
    }
    expect(presence.canSendInput(state, ASKER)).toBe(record.allowed);
  });

  it("does not let the lease holder bypass the role check in open mode", () => {
    // Open mode is not "hijack mode plus everyone else" — holding the lease
    // stops meaning anything, so a holder with no role is a viewer.
    const record = golden.input.find((entry) => entry.name === "open mode, holder with no role");
    expect(record?.is_owner).toBe(true);
    expect(record?.allowed).toBe(false);
  });

  it("treats a browser it has never seen as a viewer", () => {
    const record = golden.input.find((entry) => entry.name === "open mode, unknown browser");
    expect(record?.allowed).toBe(false);
  });
});

describe("PresenceManager.registerBrowserStateSnapshot", () => {
  it.each(golden.snapshots)("$name", async (record) => {
    const { hub, presence } = build();
    const shape = SNAPSHOT_SHAPES[record.name];
    if (shape !== undefined) {
      const state = new WorkerTermState({ now: () => NOW });
      state.workerWs = shape.connected === false ? undefined : { sendText: async () => {} };
      state.inputMode = shape.inputMode ?? "hijack";
      state.hijackOwner = shape.owner;
      state.hijackOwnerExpiresAt = shape.ownerExpiresAt;
      if (shape.restExpiresAt !== undefined) {
        state.hijackSession = { hijackId: "h1", owner: "cli", leaseExpiresAt: shape.restExpiresAt };
      }
      hub.registry.put("w1", state);
    }
    expect(await presence.registerBrowserStateSnapshot("w1", ASKER)).toStrictEqual({
      isHijacked: record.is_hijacked,
      hijackedByMe: record.hijacked_by_me,
      workerOnline: record.worker_online,
      inputMode: record.input_mode,
    });
  });

  it("answers for a worker it does not know rather than failing", async () => {
    // A browser can attach before its worker connects; an error here would
    // break the page instead of showing an empty session.
    const record = golden.snapshots.find((entry) => entry.name === "unknown worker");
    expect(record).toStrictEqual({
      name: "unknown worker",
      is_hijacked: false,
      hijacked_by_me: false,
      worker_online: false,
      input_mode: "hijack",
    });
  });

  it("does not claim a hold that has lapsed", async () => {
    const record = golden.snapshots.find((entry) => entry.name === "my hijack has expired");
    expect(record?.hijacked_by_me).toBe(false);
  });

  it("does not claim someone else's session because mine used to be here", async () => {
    // My hold lapsed and a REST client took the worker. The session is
    // hijacked and my identity is still in the owner slot, so asking the
    // broader "is anything holding this?" question here would tell the UI I
    // am still in control while someone else types.
    const record = golden.snapshots.find((entry) => entry.name === "my hold lapsed, rest took over");
    expect(record?.is_hijacked).toBe(true);
    expect(record?.hijacked_by_me).toBe(false);
  });
});

describe("PresenceManager worker pokes", () => {
  it("sends a snapshot request", async () => {
    const { hub, presence } = build();
    await presence.requestSnapshot("w1");
    expect(hub.sent).toHaveLength(1);
    expect(hub.sent[0]?.type).toBe(golden.control_frames.types[0]);
    expect(hub.sent[0]?.ts).toBe(golden.control_frames.timestamps[0]);
    expect(Object.keys(hub.sent[0] ?? {}).sort()).toStrictEqual(golden.control_frames.keys[0]);
  });

  it("sends an analysis request", async () => {
    const { hub, presence } = build();
    await presence.requestAnalysis("w1");
    expect(hub.sent[0]?.type).toBe(golden.control_frames.types[1]);
  });

  it("gives each request its own id", async () => {
    // The worker correlates its reply by this id, so two outstanding pokes
    // sharing one would have their answers confused.
    expect(golden.control_frames.req_ids_differ).toBe(true);
    const { hub, presence } = build();
    await presence.requestSnapshot("w1");
    await presence.requestSnapshot("w1");
    expect(hub.sent[0]?.req_id).not.toBe(hub.sent[1]?.req_id);
    expect(String(hub.sent[0]?.req_id).length).toBe(golden.control_frames.req_id_lengths[0]);
  });

  it("stamps a poke with wall time by default", async () => {
    const hub = new FakeHub();
    const presence = new PresenceManager({ hub });
    await presence.requestSnapshot("w1");
    expect(Math.abs(Number(hub.sent[0]?.ts) - Date.now() / 1000)).toBeLessThan(5);
  });
});

describe("PresenceManager.resolveRoleForBrowser", () => {
  it.each(golden.roles.resolved)("passes %s through from the hub's resolver", async (role) => {
    const { hub, presence } = build();
    hub.role = role;
    expect(await presence.resolveRoleForBrowser(ASKER, "w1")).toBe(role);
  });
});
